src/include/random.h

```diff
@@ -0,0 +1,391 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+/*
+ * Ceph - scalable distributed file system
+ *
+ * Copyright (C) 2017 SUSE LINUX GmbH
+ *
+ * This is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License version 2.1, as published by the Free Software
+ * Foundation.  See file COPYING.
+ *
+*/
+
+#ifndef CEPH_RANDOM_H
+#define CEPH_RANDOM_H 1
+
+#include <mutex>
+#include <random>
+
+// Basic random number facility, adapted from N3551:
+namespace ceph {
+namespace util {
+
+inline namespace version_2017_0 {
+
+namespace detail {
+
+template <typename T>
+T& urng()
+{
+ static T rng_engine;
+ return rng_engine;
+}
+
+template <typename MutexT, typename EngineT>
+void randomize_rng(const int seed, MutexT& m, EngineT& e)
+{
+ std::lock_guard<MutexT> lg(m);
+ e.seed(seed);
+}
+
+template <typename MutexT, typename EngineT>
+void randomize_rng(MutexT& m, EngineT& e)
+{
+ static std::random_device rd;
+
+ std::lock_guard<MutexT> lg(m);
+ e.seed(rd());
+}
+
+template <typename MutexT, 
+          typename EngineT = std::default_random_engine>
+void randomize_rng(MutexT& m)
+{
+ randomize_rng(m, detail::urng<EngineT>());
+}
+
+template <typename MutexT, 
+          typename EngineT = std::default_random_engine>
+void randomize_rng(const int n, MutexT& m)
+{
+ std::lock_guard<MutexT> lg(m);
+ detail::urng<EngineT>().seed(n);
+}
+
```

Comment:
I am a bit concerned about the idea of relying only on thread_local: if an engine has been provided that could be externally be pointed to across threads, we could still have contention on the engine. One idea: we can say that if the user does /not/ pass a mutex (or use, perhaps, a "locked" class) then things are thread_local. If the user /does/ pass a mutex, we lock. Doesn't seem too bad to me.

@adamemerson Let me think a while more about seeding. Your suggestion sounds pretty good, but maybe we need to leave a single randomize_rng() in-- the one with no parameters (or mutex), which will internally just use std::random_device. Then again, I do like the idea that using one of the higher-level constructs (the function generator or function object) implies a unique engine and thus avoids this situation entirely. I'll ponder a bit.

@cbodley thoughts?